vm/rust/src/recorded_state.rs

```diff
@@ -0,0 +1,123 @@
+use blockifier::context::BlockContext;
+use blockifier::execution::contract_class::ContractClass;
+
+use blockifier::state::cached_state::CachedState;
+use blockifier::state::{
+    cached_state::StorageEntry,
+    state_api::{StateReader, StateResult},
+};
+use serde::{Deserialize, Serialize};
+use starknet_api::{
+    core::{ClassHash, CompiledClassHash, ContractAddress, Nonce},
+    state::StorageKey,
+};
+use starknet_types_core::felt::Felt;
+use std::collections::HashMap;
+
+use crate::juno_state_reader::class_info_from_json_str;
+use crate::{build_block_context, execute_transaction, VMArgs};
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct RecordedState<Class> {
+    pub storage: HashMap<StorageEntry, Felt>,
+    pub nonce: HashMap<ContractAddress, Nonce>,
+    pub class_hash: HashMap<ContractAddress, ClassHash>,
+    pub contract_class: HashMap<ClassHash, Class>,
+}
+
+pub type NativeState = RecordedState<String>;
+pub type CompiledNativeState = RecordedState<ContractClass>;
+
+impl CompiledNativeState {
+    pub fn new(state: NativeState) -> Self {
+        let mut compiled_contract_class: HashMap<ClassHash, ContractClass> = Default::default();
+
+        for class_hash in state.contract_class.keys() {
+            let contract_class = state.get_compiled_contract_class(*class_hash).unwrap();
+            compiled_contract_class.insert(*class_hash, contract_class);
+        }
+
+        RecordedState {
+            storage: state.storage,
+            nonce: state.nonce,
+            class_hash: state.class_hash,
+            contract_class: compiled_contract_class,
+        }
+    }
+}
+
+trait ContractClassTrait {
+    fn compile(&self, class_hash: ClassHash) -> StateResult<ContractClass>;
+}
+
+impl ContractClassTrait for String {
+    fn compile(&self, class_hash: ClassHash) -> StateResult<ContractClass> {
+        Ok(class_info_from_json_str(self, class_hash)
+            .expect("decoding class went wrong")
+            .contract_class())
+    }
+}
+
+impl ContractClassTrait for ContractClass {
+    fn compile(&self, _class_hash: ClassHash) -> StateResult<ContractClass> {
+        Ok(self.clone())
+    }
```

Comment:
`ContractClass` doesn't have a `class_hash`. It's only there because it can be used for persisting the contract class to disk, but since we already have a contract class that isn't necessary.